src/main/java/com/github/manolo8/darkbot/core/objects/swf/FlashMap.java

```diff
@@ -0,0 +1,379 @@
+
+    private static final int kDontEnumBit = 0x01;
+    private static final int kHasDeletedItems = 0x02;
+    private static final int kHasIterIndex = 0x04;
+
+    private static final int MAX_SIZE = 1024;
+    private static final int MAX_CAPACITY = MAX_SIZE * Long.BYTES * 2;
+    private static final byte[] BUFFER = new byte[MAX_CAPACITY];
+
+    private final AtomKind keyKind;
+    private final AtomKind valueKind;
+
+    private final Class<K> keyType;
+    private final Class<V> valueType;
+
+    private final boolean keyUpdatable, valueUpdatable;
+
+    private int size;
+    private Entry[] entries;
+
+    private long address;
+    private Map<K, ValueWrapper> updatables;
+
+    public FlashMap(Class<K> keyType, Class<V> valueType) {
+        this.keyKind = AtomKind.of(keyType);
+        this.valueKind = AtomKind.of(valueType);
+
+        this.keyType = keyType;
+        this.valueType = valueType;
+        if (keyKind.isNotSupported() || valueKind.isNotSupported())
+            throw new IllegalArgumentException("Illegal java types!");
+
+        this.keyUpdatable = Updatable.class.isAssignableFrom(keyType);
+        this.valueUpdatable = Updatable.class.isAssignableFrom(valueType);
+
+        if (keyUpdatable)
+            throw new IllegalArgumentException("Key as updatable is not supported!");
+
+        clear();
+    }
+
+    private static void updateIfChanged(Updatable u, long address) {
+        if (u.address != address)
+            u.update(address);
+    }
+
+    @Override
+    public void update() {
+        if (address == 0) return;
+
+        long traits = API.readLong(address, 16, 40);
+
+        int tableOffset = API.readInt(traits, 236);
+        if (tableOffset == 0) return;
+
+        boolean isDictionary = (API.readInt(traits, 248) & (1 << 4)) != 0;
+        if (isDictionary) {
+            readHashTable(API.readLong(address + tableOffset) + 8); //read hash table ptr & skip cpp vtable
+        } else {
+            readHashTable(address + tableOffset);
+        }
+    }
+
+    @Override
+    public void update(long address) {
+        if (this.address != address) {
+            clear();
+        }
+
+        this.address = address;
+    }
+
+    private int getCapacity(int logCapacity, @SuppressWarnings("unused") boolean hasIterIndexes) {
+        if (logCapacity <= 0) return 0;
+
+        logCapacity = 1 << (logCapacity - 1);
+        //if (hasIterIndexes) logCapacity += 2; // for tests only
+
+        return logCapacity * Long.BYTES;
+    }
+
+    private void readHashTable(long table) {
+        long atomsAndFlags = API.readLong(table);
+        long atoms = (atomsAndFlags & ByteUtils.ATOM_MASK) + 8;
+
+        int size = API.readMemoryInt(table + 8); // includes deleted items
+        int capacity = getCapacity(API.readMemoryInt(table + 12), (atomsAndFlags & kHasIterIndex) != 0);
+
+        if (size <= 0 || size > MAX_SIZE || capacity <= 0 || capacity > MAX_CAPACITY) {
```

Comment:
is this max size (of 1024) an actual hard-limit? i know in the past we've had to raise it on arrays, since players got over 512 (old limit) entities on the map and it would stop updating all together. we raised it to 8*1024 in that case, unsure about this use-case tho